### PR TITLE
Fix missing icon props in button `styleFrom`  methods

### LINKED ISSUE #154798: Add missing icon props in button `styleFrom`  methods.
### Steps to reproduce
Following button `styleFrom` methods are missing some icon props:

- `ElevatedButton.styleFrom` (missing `iconSize`)
- `FilledButton.styleFrom` (missing `iconSize`)
- `OutlinedButton.styleFrom` (missing `iconSize`)
- `TextButton.styleFrom` (missing `iconSize`)
- `MenuItemButton.styleFrom` (missing `iconSize` and `disabledIconColor`)
- `SubmenuButton.styleFrom` (missing `iconSize` and `disabledIconColor`)
- `SegmentedButton.styleFrom` (missing `iconSize`, `iconColor`, and `disabledIconColor`)

### Expected results

Can access all icon propers from regular `ButtonStyle` in `styleFrom` methods.

### Actual results

Missing some icon props

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

enum Calendar { day, week, month, year }

void main() => runApp(const MyApp());

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  Calendar calendarView = Calendar.week;
  bool isEnabled = true;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: Center(
          child: Column(
            spacing: 10.0,
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              ElevatedButton.icon(
                style: ElevatedButton.styleFrom(
                  iconSize: 30,
                  iconColor: Colors.red,
                  disabledIconColor: Colors.red.withValues(alpha: 0.5),
                ),
                onPressed: isEnabled ? () {} : null,
                icon: const Icon(Icons.add),
                label: const Text('ElevatedButton'),
              ),
              FilledButton.icon(
                style: ElevatedButton.styleFrom(
                  iconSize: 30,
                  iconColor: Colors.red,
                  disabledIconColor: Colors.red.withValues(alpha: 0.5),
                ),
                onPressed: isEnabled ? () {} : null,
                icon: const Icon(Icons.add),
                label: const Text('FilledButton'),
              ),
              FilledButton.tonalIcon(
                style: ElevatedButton.styleFrom(
                  iconSize: 30,
                  iconColor: Colors.red,
                  disabledIconColor: Colors.red.withValues(alpha: 0.5),
                ),
                onPressed: isEnabled ? () {} : null,
                icon: const Icon(Icons.add),
                label: const Text('Add'),
              ),
              OutlinedButton.icon(
                style: ElevatedButton.styleFrom(
                  iconSize: 30,
                  iconColor: Colors.red,
                  disabledIconColor: Colors.red.withValues(alpha: 0.5),
                ),
                onPressed: isEnabled ? () {} : null,
                icon: const Icon(Icons.add),
                label: const Text('OutlinedButton'),
              ),
              TextButton.icon(
                style: ElevatedButton.styleFrom(
                  iconSize: 30,
                  iconColor: Colors.red,
                  disabledIconColor: Colors.red.withValues(alpha: 0.5),
                ),
                onPressed: isEnabled ? () {} : null,
                icon: const Icon(Icons.add),
                label: const Text('TextButton'),
              ),
              SizedBox(
                width: 200,
                child: MenuItemButton(
                  style: MenuItemButton.styleFrom(
                    iconSize: 30,
                    iconColor: Colors.red,
                    disabledIconColor: Colors.red.withValues(alpha: 0.5),
                  ),
                  trailingIcon: const Icon(Icons.arrow_forward_ios),
                  onPressed: isEnabled ? () {} : null,
                  child: const Text('MenuItemButton'),
                ),
              ),
              SizedBox(
                width: 200,
                child: SubmenuButton(
                  style: SubmenuButton.styleFrom(
                    iconSize: 30,
                    iconColor: Colors.red,
                    disabledIconColor: Colors.red.withValues(alpha: 0.5),
                  ),
                  trailingIcon: const Icon(Icons.arrow_forward_ios),
                  menuChildren: <Widget>[
                    if (isEnabled) const Text('Item'),
                  ],
                  child: const Text('SubmenuButton'),
                ),
              ),
              SegmentedButton<Calendar>(
                style: SegmentedButton.styleFrom(
                  iconColor: Colors.red,
                  iconSize: 30,
                  disabledIconColor: Colors.red.withValues(alpha: 0.5),
                ),
                segments: const <ButtonSegment<Calendar>>[
                  ButtonSegment<Calendar>(
                      value: Calendar.day,
                      label: Text('Day'),
                      icon: Icon(Icons.calendar_view_day)),
                  ButtonSegment<Calendar>(
                      value: Calendar.week,
                      label: Text('Week'),
                      icon: Icon(Icons.calendar_view_week)),
                  ButtonSegment<Calendar>(
                      value: Calendar.month,
                      label: Text('Month'),
                      icon: Icon(Icons.calendar_view_month)),
                  ButtonSegment<Calendar>(
                      value: Calendar.year,
                      label: Text('Year'),
                      icon: Icon(Icons.calendar_today)),
                ],
                selected: <Calendar>{calendarView},
                onSelectionChanged:
                    isEnabled ? (Set<Calendar> newSelection) {} : null,
              )
            ],
          ),
        ),
        floatingActionButton: FloatingActionButton.extended(
          onPressed: () {
            setState(() {
              isEnabled = !isEnabled;
            });
          },
          label: Text(isEnabled ? 'Enabled' : 'Disabled'),
        ),
      ),
    );
  }
}
```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[!] Flutter (Channel main, 3.25.0-1.0.pre.283, on macOS 15.0 24A5327a
    darwin-arm64, locale en-US)
    • Flutter version 3.25.0-1.0.pre.283 on channel main at
      /Users/tahatesser/Code/flutter
    ! Upstream repository git@github.com:TahaTesser/flutter.git is not the same
      as FLUTTER_GIT_URL
    • FLUTTER_GIT_URL = git@github.com:TahaTesser/flutter.gitexport
    • Framework revision aea84342eb (3 days ago), 2024-09-05 16:46:22 -0700
    • Engine revision 015f3b1dec
    • Dart version 3.6.0 (build 3.6.0-216.0.dev)
    • DevTools version 2.39.0
    • If those were intentional, you can disregard the above warnings; however
      it is recommended to use "git" directly to perform update checks and
      upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 35.0.0)
    • Android SDK at /Users/tahatesser/Code/android-sdk
    • Platform android-35, build-tools 35.0.0
    • ANDROID_SDK_ROOT = /Users/tahatesser/Code/android-sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      17.0.11+0-17.0.11b1207.24-11852314)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 16.0)
    • Xcode at /Applications/Xcode-beta.app/Contents/Developer
    • Build 16A5230g
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2024.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      17.0.11+0-17.0.11b1207.24-11852314)

[✓] VS Code (version 1.92.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.94.0

[✓] Connected device (4 available)
    • Taha’s iPhone (mobile)          • 00008120-001C559034DB401E • ios
      • iOS 18.0 22A5350a
    • macOS (desktop)                 • macos                     • darwin-arm64
      • macOS 15.0 24A5327a darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad     • darwin
      • macOS 15.0 24A5327a darwin-arm64
    • Chrome (web)                    • chrome                    •
      web-javascript • Google Chrome 128.0.6613.120

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.

```

</details>

### PR DESCRIPTION
**Pull request by TahaTesser**
Fixes [Add missing icon props in button `styleFrom`  methods.](https://github.com/flutter/flutter/issues/154798)

### Description

Add missing icon propers in the following widgets:

- `ElevatedButton.styleFrom` (missing `iconSize`)
- `FilledButton.styleFrom` (missing `iconSize`)
- `OutlinedButton.styleFrom` (missing `iconSize`)
- `TextButton.styleFrom` (missing `iconSize`)
- `MenuItemButton.styleFrom` (missing `iconSize` and `disabledIconColor`)
- `SubmenuButton.styleFrom` (missing `iconSize` and `disabledIconColor`)
- `SegmentedButton.styleFrom` (missing `iconSize`, `iconColor`, and `disabledIconColor`)

### Code sample

<details>
<summary>expand to view the code sample</summary> 

```dart
import 'package:flutter/material.dart';

enum Calendar { day, week, month, year }

void main() => runApp(const MyApp());

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  Calendar calendarView = Calendar.week;
  bool isEnabled = true;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: Center(
          child: Column(
            spacing: 10.0,
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              ElevatedButton.icon(
                style: ElevatedButton.styleFrom(
                  iconSize: 30,
                  iconColor: Colors.red,
                  disabledIconColor: Colors.red.withValues(alpha: 0.5),
                ),
                onPressed: isEnabled ? () {} : null,
                icon: const Icon(Icons.add),
                label: const Text('ElevatedButton'),
              ),
              FilledButton.icon(
                style: ElevatedButton.styleFrom(
                  iconSize: 30,
                  iconColor: Colors.red,
                  disabledIconColor: Colors.red.withValues(alpha: 0.5),
                ),
                onPressed: isEnabled ? () {} : null,
                icon: const Icon(Icons.add),
                label: const Text('FilledButton'),
              ),
              FilledButton.tonalIcon(
                style: ElevatedButton.styleFrom(
                  iconSize: 30,
                  iconColor: Colors.red,
                  disabledIconColor: Colors.red.withValues(alpha: 0.5),
                ),
                onPressed: isEnabled ? () {} : null,
                icon: const Icon(Icons.add),
                label: const Text('Add'),
              ),
              OutlinedButton.icon(
                style: ElevatedButton.styleFrom(
                  iconSize: 30,
                  iconColor: Colors.red,
                  disabledIconColor: Colors.red.withValues(alpha: 0.5),
                ),
                onPressed: isEnabled ? () {} : null,
                icon: const Icon(Icons.add),
                label: const Text('OutlinedButton'),
              ),
              TextButton.icon(
                style: ElevatedButton.styleFrom(
                  iconSize: 30,
                  iconColor: Colors.red,
                  disabledIconColor: Colors.red.withValues(alpha: 0.5),
                ),
                onPressed: isEnabled ? () {} : null,
                icon: const Icon(Icons.add),
                label: const Text('TextButton'),
              ),
              SizedBox(
                width: 200,
                child: MenuItemButton(
                  style: MenuItemButton.styleFrom(
                    iconSize: 30,
                    iconColor: Colors.red,
                    disabledIconColor: Colors.red.withValues(alpha: 0.5),
                  ),
                  trailingIcon: const Icon(Icons.arrow_forward_ios),
                  onPressed: isEnabled ? () {} : null,
                  child: const Text('MenuItemButton'),
                ),
              ),
              SizedBox(
                width: 200,
                child: SubmenuButton(
                  style: SubmenuButton.styleFrom(
                    iconSize: 30,
                    iconColor: Colors.red,
                    disabledIconColor: Colors.red.withValues(alpha: 0.5),
                  ),
                  trailingIcon: const Icon(Icons.arrow_forward_ios),
                  menuChildren: <Widget>[
                    if (isEnabled) const Text('Item'),
                  ],
                  child: const Text('SubmenuButton'),
                ),
              ),
              SegmentedButton<Calendar>(
                style: SegmentedButton.styleFrom(
                  iconColor: Colors.red,
                  iconSize: 30,
                  disabledIconColor: Colors.red.withValues(alpha: 0.5),
                ),
                segments: const <ButtonSegment<Calendar>>[
                  ButtonSegment<Calendar>(
                      value: Calendar.day,
                      label: Text('Day'),
                      icon: Icon(Icons.calendar_view_day)),
                  ButtonSegment<Calendar>(
                      value: Calendar.week,
                      label: Text('Week'),
                      icon: Icon(Icons.calendar_view_week)),
                  ButtonSegment<Calendar>(
                      value: Calendar.month,
                      label: Text('Month'),
                      icon: Icon(Icons.calendar_view_month)),
                  ButtonSegment<Calendar>(
                      value: Calendar.year,
                      label: Text('Year'),
                      icon: Icon(Icons.calendar_today)),
                ],
                selected: <Calendar>{calendarView},
                onSelectionChanged:
                    isEnabled ? (Set<Calendar> newSelection) {} : null,
              )
            ],
          ),
        ),
        floatingActionButton: FloatingActionButton.extended(
          onPressed: () {
            setState(() {
              isEnabled = !isEnabled;
            });
          },
          label: Text(isEnabled ? 'Enabled' : 'Disabled'),
        ),
      ),
    );
  }
}
```

</details>

### Preview (Customized using icon props in `styleFrom`  methods)

Screenshot 2024-09-09 at 16 27 19

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter/test/material/elevated_button_test.dart [hunk 1]
```diff
@@ -2400,11 +2406,103 @@ void main() {
...
+    expect(tester.getSize(find.byIcon(Icons.add)), const Size(iconSize, iconSize));
+    expect(iconStyle(tester, Icons.add).color, iconColor);
+
+    // Test disabled button.
+    await tester.pumpWidget(buildButton(enabled: false));
+    expect(iconStyle(tester, Icons.add).color, disabledIconColor);
+  });
+
+  // This is a test regression for https://github.com/flutter/flutter/issues/154798.
+  test('ElevatedButton.styleFrom constructor', () {
```
Comment (QuncCccccc) [inline packages/flutter/test/material/elevated_button_test.dart:2443]: Just a question, what is this unit test testing for?

File: packages/flutter/lib/src/material/elevated_button.dart [hunk 2]
```diff
@@ -156,9 +156,10 @@ class ElevatedButton extends ButtonStyleButton {
   /// used to create a [WidgetStateProperty] [ButtonStyle.backgroundColor].
   ///
   /// Similarly, the [enabledMouseCursor] and [disabledMouseCursor]
-  /// parameters are used to construct [ButtonStyle.mouseCursor] and
+  /// parameters are used to construct [ButtonStyle.mouseCursor], and
   /// [iconColor], [disabledIconColor] are used to construct
-  /// [ButtonStyle.iconColor].
+  /// [ButtonStyle.iconColor], and [iconSize] is used to construct
+  /// [ButtonStyle.iconSize].
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/elevated_button.dart:162]: Tiny nit: the triple-compound sentence is a little bit hard to follow.

I noticed that the other `ButtonStyleButton.allOrNull()` values aren't included in this documentation. Perhaps they're simple enough that no docs are necessary, or maybe they could all be included.

File: packages/flutter/test/material/elevated_button_test.dart (hunk 1 above)
Comment (TahaTesser) [reply packages/flutter/test/material/elevated_button_test.dart:2443]: This tests all the expected styleFrom method parameters. Some existing tests only some of the styleFrom but not every parameter hence we missed some icon properties.

This test is regular test not a widget tests just to props similar to debug fill properties tests.

File: packages/flutter/test/material/elevated_button_test.dart [hunk 3]
```diff
@@ -10,6 +10,13 @@ import 'package:flutter_test/flutter_test.dart';
 import '../widgets/semantics_tester.dart';
 
 void main() {
+  TextStyle iconStyle(WidgetTester tester, IconData icon) {
+    final RichText iconRichText = tester.widget<RichText>(
+      find.descendant(of: find.byIcon(icon), matching: find.byType(RichText)),
+    );
+    return iconRichText.text.style!;
+  }
```
Comment (nate-thegrate) [inline packages/flutter/test/material/elevated_button_test.dart:18]: As far as "global functions" vs. "functions inside main", I'm not sure if one is better than the other. The style guide has a section called [**Make each test entirely self-contained**](https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#make-each-test-entirely-self-contained), but the goal of the rule is to avoid shared state, so a helper function like this one should be good 👍

File: packages/flutter/test/material/elevated_button_test.dart [hunk 4]
```diff
@@ -2400,11 +2406,103 @@ void main() {
...
     expect(buttonTopLeft.dx, iconTopLeft.dx - 24.0); // 24.0 - padding between icon and button edge.
   });
-}
 
-TextStyle _iconStyle(WidgetTester tester, IconData icon) {
-  final RichText iconRichText = tester.widget<RichText>(
-    find.descendant(of: find.byIcon(icon), matching: find.byType(RichText)),
-  );
-  return iconRichText.text.style!;
+  // This is a test regression for https://github.com/flutter/flutter/issues/154798.
```
Comment (nate-thegrate) [inline packages/flutter/test/material/elevated_button_test.dart:2410]:
```suggestion
  // Regression test for https://github.com/flutter/flutter/issues/154798.
```

File: packages/flutter/test/material/elevated_button_test.dart [hunk 5]
```diff
@@ -2400,11 +2406,103 @@ void main() {
...
+    const Duration animationDuration = Duration(milliseconds: 200);
+    const bool enableFeedback = true;
+    const AlignmentGeometry alignment = Alignment.center;
+    const InteractiveInkFeatureFactory splashFactory = InkRipple.splashFactory;
+    Widget backgroundBuilder(BuildContext context, Set<MaterialState> states, Widget? child) {
+      return const SizedBox();
+    }
+    Widget foregroundBuilder(BuildContext context, Set<MaterialState> states, Widget? child) {
+      return const SizedBox();
+    }
```
Comment (nate-thegrate) [inline packages/flutter/test/material/elevated_button_test.dart:2475]: We should probably avoid introducing new references to deprecated names, so maybe `Set<WidgetState>` would be the ideal parameter type.

File: packages/flutter/test/material/segmented_button_test.dart [hunk 6]
```diff
@@ -1122,6 +1129,111 @@ void main() {
...
+    const Duration animationDuration = Duration(milliseconds: 200);
+    const bool enableFeedback = true;
+    const AlignmentGeometry alignment = Alignment.center;
+    const InteractiveInkFeatureFactory splashFactory = InkRipple.splashFactory;
+
+    SegmentedButton.styleFrom(
+      backgroundColor: backgroundColor,
+      foregroundColor: foregroundColor,
+      disabledForegroundColor: disabledForegroundColor,
+      disabledBackgroundColor: disabledBackgroundColor,
```
Comment (nate-thegrate) [inline packages/flutter/test/material/segmented_button_test.dart:1212]: Definitely not a huge deal, but one way to make this more concise would be to do

```dart
    SegmentedButton.styleFrom(
      foregroundColor: const Color(0xFF000022),
      backgroundColor: const Color(0xFF000011),
      disabledForegroundColor: const Color(0xFF000033),
      disabledBackgroundColor: const Color(0xFF000044),
      // ...
    );
```

It could make it easier to prevent confusion caused by e.g. `foregroundColor` and `backgroundColor` being defined in the opposite order of where they're referenced.

**Review by nate-thegrate — COMMENTED** [review]
This looks really great overall. I wanted to share a few tiny suggestions (and unfortunately they involve a lot of copy/pasting, so apologies for that).

File: packages/flutter/test/material/elevated_button_test.dart (hunk 1 above)
Comment (justinmc) [reply packages/flutter/test/material/elevated_button_test.dart:2443]: What if you create an additional ButtonStyle with the same parameters but with the normal ButtonStyle constructor, and then compare their equality? Not sure if that would work like I'm thinking it would, but if so it would fail if an additional parameter is added to ButtonStyle but not to styleFrom, which would be pretty valuable.

**Comment by nate-thegrate** [comment]
Also, apologies for the merge conflicts, that's on me 😅

(though to be fair, https://github.com/flutter/flutter/pull/154695 was opened before https://github.com/flutter/flutter/issues/154798 was)

File: packages/flutter/lib/src/material/elevated_button.dart (hunk 2 above)
Comment (TahaTesser) [reply packages/flutter/lib/src/material/elevated_button.dart:162]: `iconColor` docs are included in some of the docs, if not all values. I'll split the icon related sentences form cursor.

File: packages/flutter/test/material/elevated_button_test.dart (hunk 3 above)
Comment (TahaTesser) [reply packages/flutter/test/material/elevated_button_test.dart:18]: Generally we put functions inside main.

File: packages/flutter/test/material/elevated_button_test.dart (hunk 1 above)
Comment (TahaTesser) [reply packages/flutter/test/material/elevated_button_test.dart:2443]: @justinmc  Something like this?

```dart
  // Regression for https://github.com/flutter/flutter/issues/154798.
  test('ElevatedButton.styleFrom constructor', () {
    const Color foregroundColor = Color(0xFF000022);
    const Color backgroundColor = Color(0xFF000011);
    const Color disabledForegroundColor = Color(0xFF000033);
    const Color disabledBackgroundColor = Color(0xFF000044);
    const Color shadowColor = Color(0xFF000066);
    const Color surfaceTintColor = Color(0xFF000077);
    const Color iconColor = Color(0xFF000088);
    const double iconSize = 32.0;
    const Color disabledIconColor = Color(0xFF000099);
    const Color overlayColor = Color(0xFF000055);
    const double elevation = 10.0;
    const TextStyle textStyle = TextStyle(fontSize: 21.0);
    const EdgeInsetsGeometry padding = EdgeInsets.all(16.0);
    const Size minimumSize = Size(88.0, 36.0);
    const Size fixedSize = Size(100.0, 50.0);
    const Size maximumSize = Size(120.0, 48.0);
    const BorderSide side = BorderSide(width: 2.0, color: Color(0xFF0000AA));
    final OutlinedBorder shape = RoundedRectangleBorder(borderRadius: BorderRadius.circular(8.0));
    const MouseCursor enabledMouseCursor = SystemMouseCursors.click;
    const MouseCursor disabledMouseCursor = SystemMouseCursors.basic;
    const VisualDensity visualDensity = VisualDensity.compact;
    const MaterialTapTargetSize tapTargetSize = MaterialTapTargetSize.shrinkWrap;
    const Duration animationDuration = Duration(milliseconds: 200);
    const bool enableFeedback = true;
    const AlignmentGeometry alignment = Alignment.center;
    const InteractiveInkFeatureFactory splashFactory = InkRipple.splashFactory;
    Widget backgroundBuilder(BuildContext context, Set<WidgetState> states, Widget? child) {
      return const SizedBox();
    }
    Widget foregroundBuilder(BuildContext context, Set<WidgetState> states, Widget? child) {
      return const SizedBox();
    }

    final ButtonStyle styleFrom = ElevatedButton.styleFrom(
      foregroundColor: foregroundColor,
      backgroundColor: backgroundColor,
      disabledForegroundColor: disabledForegroundColor,
      disabledBackgroundColor: disabledBackgroundColor,
      shadowColor: shadowColor,
      surfaceTintColor: surfaceTintColor,
      iconColor: iconColor,
      iconSize: iconSize,
      disabledIconColor: disabledIconColor,
      overlayColor: overlayColor,
      elevation: elevation,
      textStyle: textStyle,
      padding: padding,
      minimumSize: minimumSize,
      fixedSize: fixedSize,
      maximumSize: maximumSize,
      side: side,
      shape: shape,
      enabledMouseCursor: enabledMouseCursor,
      disabledMouseCursor: disabledMouseCursor,
      visualDensity: visualDensity,
      tapTargetSize: tapTargetSize,
      animationDuration: animationDuration,
      enableFeedback: enableFeedback,
      alignment: alignment,
      splashFactory: splashFactory,
      backgroundBuilder: backgroundBuilder,
      foregroundBuilder: foregroundBuilder,
    );

    final ButtonStyle buttonStyle = ButtonStyle(
      foregroundColor: const WidgetStateProperty<Color?>.fromMap(
        <WidgetStatesConstraint, Color?>{
          WidgetState.disabled: disabledForegroundColor,
          WidgetState.any: foregroundColor,
        },
      ),
      backgroundColor: const WidgetStateProperty<Color?>.fromMap(
        <WidgetStatesConstraint, Color?>{
          WidgetState.disabled: disabledBackgroundColor,
          WidgetState.any: backgroundColor,
        },
      ),
      shadowColor: const WidgetStatePropertyAll<Color?>(shadowColor),
      surfaceTintColor: const WidgetStatePropertyAll<Color?>(surfaceTintColor),
      iconColor: const WidgetStateProperty<Color?>.fromMap(
        <WidgetStatesConstraint, Color?>{
          WidgetState.disabled: disabledIconColor,
          WidgetState.any: iconColor,
        },
      ),
      iconSize: const WidgetStatePropertyAll<double>(iconSize),
      overlayColor: const WidgetStatePropertyAll<Color?>(overlayColor),
      elevation: const WidgetStatePropertyAll<double>(elevation),
      textStyle: const WidgetStatePropertyAll<TextStyle?>(textStyle),
      padding: const WidgetStatePropertyAll<EdgeInsetsGeometry>(padding),
      minimumSize: const WidgetStatePropertyAll<Size>(minimumSize),
      fixedSize: const WidgetStatePropertyAll<Size>(fixedSize),
      maximumSize: const WidgetStatePropertyAll<Size>(maximumSize),
      side: const WidgetStatePropertyAll<BorderSide>(side),
      shape: WidgetStatePropertyAll<OutlinedBorder>(shape),
      mouseCursor: const WidgetStateProperty<MouseCursor?>.fromMap(
        <WidgetStatesConstraint, MouseCursor?>{
          WidgetState.disabled: disabledMouseCursor,
          WidgetState.any: enabledMouseCursor,
        },
      ),
      visualDensity: visualDensity,
      tapTargetSize: tapTargetSize,
      animationDuration: animationDuration,
      enableFeedback: enableFeedback,
      alignment: alignment,
      splashFactory: splashFactory,
      backgroundBuilder: backgroundBuilder,
      foregroundBuilder: foregroundBuilder,
    );

    expect(
      styleFrom.foregroundColor!.resolve(<WidgetState>{ WidgetState.disabled }),
      buttonStyle.foregroundColor!.resolve(<WidgetState>{ WidgetState.disabled }),
    );
    expect(
      styleFrom.foregroundColor!.resolve(<WidgetState>{}),
      buttonStyle.foregroundColor!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.backgroundColor!.resolve(<WidgetState>{ WidgetState.disabled }),
      buttonStyle.backgroundColor!.resolve(<WidgetState>{ WidgetState.disabled }),
    );
    expect(
      styleFrom.backgroundColor!.resolve(<WidgetState>{}),
      buttonStyle.backgroundColor!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.shadowColor!.resolve(<WidgetState>{}),
      buttonStyle.shadowColor!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.surfaceTintColor!.resolve(<WidgetState>{}),
      buttonStyle.surfaceTintColor!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.iconColor!.resolve(<WidgetState>{ WidgetState.disabled }),
      buttonStyle.iconColor!.resolve(<WidgetState>{ WidgetState.disabled }),
    );
    expect(
      styleFrom.iconColor!.resolve(<WidgetState>{}),
      buttonStyle.iconColor!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.iconSize!.resolve(<WidgetState>{}),
      buttonStyle.iconSize!.resolve(<WidgetState>{}),
    );
    expect(styleFrom.overlayColor!.resolve(<WidgetState>{}), null); // This is null when a foreground is provided.
    expect(
      styleFrom.elevation!.resolve(<WidgetState>{}),
      buttonStyle.elevation!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.textStyle!.resolve(<WidgetState>{}),
      buttonStyle.textStyle!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.padding!.resolve(<WidgetState>{}),
      buttonStyle.padding!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.minimumSize!.resolve(<WidgetState>{}),
      buttonStyle.minimumSize!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.fixedSize!.resolve(<WidgetState>{}),
      buttonStyle.fixedSize!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.maximumSize!.resolve(<WidgetState>{}),
      buttonStyle.maximumSize!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.side!.resolve(<WidgetState>{}),
      buttonStyle.side!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.shape!.resolve(<WidgetState>{}),
      buttonStyle.shape!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.mouseCursor!.resolve(<WidgetState>{ WidgetState.disabled }),
      buttonStyle.mouseCursor!.resolve(<WidgetState>{ WidgetState.disabled }),
    );
    expect(
      styleFrom.mouseCursor!.resolve(<WidgetState>{}),
      buttonStyle.mouseCursor!.resolve(<WidgetState>{}),
    );
    expect(styleFrom.visualDensity, buttonStyle.visualDensity);
    expect(styleFrom.tapTargetSize, buttonStyle.tapTargetSize);
    expect(styleFrom.animationDuration, buttonStyle.animationDuration);
    expect(styleFrom.enableFeedback, buttonStyle.enableFeedback);
    expect(styleFrom.alignment, buttonStyle.alignment);
    expect(styleFrom.splashFactory, buttonStyle.splashFactory);
    expect(styleFrom.backgroundBuilder, buttonStyle.backgroundBuilder);
    expect(styleFrom.foregroundBuilder, buttonStyle.foregroundBuilder);
  });
```

**Comment by TahaTesser** [comment]
> Also, apologies for the merge conflicts, that's on me 😅
> 
> (though to be fair, #154695 was opened before #154798 was)

Interesting that I see no conflict in text_button.dart, only in other buttons.

https://github.com/flutter/flutter/blob/d6e2b6aedbb7d22ddcd950e3634efd285368867f/packages/flutter/lib/src/material/text_button.dart#L252

https://github.com/flutter/flutter/blob/d6e2b6aedbb7d22ddcd950e3634efd285368867f/packages/flutter/lib/src/material/elevated_button.dart#L263

**Comment by TahaTesser** [comment]
> Also, apologies for the merge conflicts, that's on me 😅
> 
> (though to be fair, #154695 was opened before #154798 was)

> Also, apologies for the merge conflicts, that's on me 😅
> 
> (though to be fair, #154695 was opened before #154798 was)

FYI, I see no conflict in `text_button.dart`, only in other buttons. The `iconColor` line isn't updated in this class.

https://github.com/flutter/flutter/blob/d6e2b6aedbb7d22ddcd950e3634efd285368867f/packages/flutter/lib/src/material/text_button.dart#L252

https://github.com/flutter/flutter/blob/d6e2b6aedbb7d22ddcd950e3634efd285368867f/packages/flutter/lib/src/material/elevated_button.dart#L263

File: packages/flutter/test/material/elevated_button_test.dart (hunk 1 above)
Comment (nate-thegrate) [reply packages/flutter/test/material/elevated_button_test.dart:2443]: My guess is that it would just need 1 equality check:

```dart
expect(styleFrom == buttonStyle, isTrue);
```

Though if we were to add a `ButtonStyle` property without updating this test, they would both be initialized to `null` and we likely wouldn't be able to catch that `.styleFrom()` was missing something.

**Review by nate-thegrate — APPROVED** [review]
LGTM!

There's a rich variety of improvements in this PR; looking forward to seeing it get merged in 🙂

**Comment by TahaTesser** [comment]
> expect(styleFrom == buttonStyle, isTrue);

`expect(styleFrom == buttonStyle, isTrue);` is same as `expect(styleFrom == buttonStyle, isTrue);`. 

I tried this already. Either checks won't work  as the some of the individuals properties return types. We will have to break down the each property like i did in the https://github.com/flutter/flutter/pull/154821#discussion_r1751495167.

File: packages/flutter/test/material/elevated_button_test.dart (hunk 1 above)
Comment (TahaTesser) [reply packages/flutter/test/material/elevated_button_test.dart:2443]: `expect(styleFrom == buttonStyle, isTrue);` is same as `expect(styleFrom, buttonStyle);`. 

I tried this already. Either checks won't work  as the some of the individuals properties return different types. We will have to break down each properties like i did in the https://github.com/flutter/flutter/pull/154821#discussion_r1751495167.

**Review by QuncCccccc — APPROVED** [review]
LGTM:) Thanks!
